File: Adapter.cpp
#include <algorithm>
#include <iostream>

class Target { // Classe Alvo
 public:
  virtual ~Target() = default;

  virtual std::string Request() const {
    return "Target: The default target's behavior. / Alvo: O comportamento padrão do alvo.";
  }
};

class Adaptee { // Classe incompatível.
 public:
  std::string SpecificRequest() const {
    return ".eetpadA ed laicepse otnematropmoC / .eetpadA eht fo roivaheb laicepS";
  }
};

class Adapter : public Target { // Classe Adaptador.
 private:
  Adaptee *adaptee_;

 public:
  Adapter(Adaptee *adaptee) : adaptee_(adaptee) {}
  std::string Request() const override {
    std::string to_reverse = this->adaptee_->SpecificRequest();
    std::reverse(to_reverse.begin(), to_reverse.end());
    return "Adapter: (TRANSLATED) " + to_reverse;
  }
};

void ClientCode(const Target *target) { // Código do cliente.
  std::cout << target->Request();
}

int main() {
  Target *target = new Target;
  std::cout << "Client: I can work just fine with the Target objects: / Cliente: Eu consigo trabalhar bem com objetos Alvo: \n";
  ClientCode(target);
  std::cout << "\n\n";
    
  Adaptee *adaptee = new Adaptee;
  std::cout << "Client: The Adaptee class has a weird interface. See, I don't understand it: / Cliente: A classe Adaptee possui uma interface estranha. Não consigo entender: \n";
  std::cout << "Adaptee: " << adaptee->SpecificRequest();
  std::cout << "\n\n";
    
  Adapter *adapter = new Adapter(adaptee);
  std::cout << "Client: But I can work with it via the Adapter: / Cliente: Mas eu consigo trabalhar via Adaptador: \n";
  ClientCode(adapter);
  std::cout << "\n";

  delete target;
  delete adaptee;
  delete adapter;

  return 0;
}

File: 01_adapter/Adapter.py
class Target:
    """
    The Target defines the domain-specific interface used by the client code.

    A classe Alvo define a interface específica de domínio usada pelo código do cliente.
    """

    def request(self) -> str:
        return "Target: The default target's behavior. / Alvo: O comportamento padrão da classe Alvo."


class Adaptee:
    """
    The Adaptee contains some useful behavior, but its interface is incompatible
    with the existing client code. The Adaptee needs some adaptation before the
    client code can use it.

    A classe Adaptee contém algum comportamento útil, mas a interface é incompatível com o código do
    cliente existente. A classe Adaptee precisa de alguma adaptação antes do código do cliente ser
    usado.
    """

    def specific_request(self) -> str:
        return ".eetpadA eht fo roivaheb laicepS / Comportamento especial da classe Adaptee."


class Adapter(Target, Adaptee):
    """
    The Adapter makes the Adaptee's interface compatible with the Target's
    interface via multiple inheritance.

    A classe Adapter faz com que a interace da classe Adaptee seja compotível com a interface da classe
    Target via herança múltipla.
    """

    def request(self) -> str:
        return f"Adapter: (TRANSLATED / TRADUZIDO) {self.specific_request()[::-1]}"

 
def client_code(target: "Target") -> None:
    """
    The client code supports all classes that follow the Target interface.

    O código do cliente suporta todas classes que seguem a interface Alvo.
    """

    print(target.request(), end="")


if __name__ == "__main__":
    print("Client: I can work just fine with the Target objects: / Cliente: Eu funciono bem com os objetos Target")
    target = Target()
    client_code(target)
    print("\n")

    adaptee = Adaptee()
    print("Client: The Adaptee class has a weird interface.  / Cliente: A classe Adaptee tem uma interface estranha."
          "See, I don't understand it: / Veja, eu não consigo entender isto: ")
    print(f"Adaptee: {adaptee.specific_request()}", end="\n\n")

    print("Client: But I can work with it via the Adapter: / Cliente: Mas eu consigo trabalhar com isso via o Adapter")
    adapter = Adapter()
    client_code(adapter)